services/updates-svc/internal/updates/grpc.go

```diff
@@ -0,0 +1,133 @@
+package updates
+
+import (
+	"common/auth"
+	"context"
+	pb "gen/services/updates_svc/v1"
+	"github.com/EventStore/EventStore-Client-Go/v4/esdb"
+	"hwes"
+	"telemetry"
+	"time"
+)
+
+type UpdatesGrpcService struct {
+	pb.UnimplementedUpdatesServiceServer
+	es *esdb.Client
+}
+
+func NewUpdatesGrpcService(es *esdb.Client) *UpdatesGrpcService {
+	return &UpdatesGrpcService{
+		es: es,
+	}
+}
+
+func (s *UpdatesGrpcService) ReceiveUpdates(req *pb.ReceiveUpdatesRequest, stream pb.UpdatesService_ReceiveUpdatesServer) error {
+	ctx, span, log := telemetry.StartSpan(stream.Context(), "updates.ReceiveUpdates")
+	defer span.End()
+
+	tokenExpires, err := auth.SessionValidUntil(ctx)
+	if err != nil {
+		return err
+	}
+
+	tokenExpiresIn := time.Until(tokenExpires)
+
+	ctx, cancel := context.WithCancel(ctx)
+	// Close stream after session is invalid
+	time.AfterFunc(tokenExpiresIn, func() {
+		log.Debug().Msg("closing stream, token expired")
+		cancel()
+	})
+
+	log.Debug().
+		// We are using .Str() instead of .Dur() to enhance human readability in the logs
+		// .Dur() -> 210000
+		// .Str() -> 3m30s
+		Str("closeStreamIn", tokenExpiresIn.String()).
+		Msg("will close session when token expires")
+
+	organizationID, err := auth.GetOrganizationID(ctx)
+	if err != nil {
+		return err
+	}
+
+	esSubscribeToAllOptions := esdb.SubscribeToAllOptions{
+		From:   esdb.End{},
+		Filter: esdb.ExcludeSystemEventsFilter(),
+	}
+
+	if req.Revision != nil {
+		log.Debug().
+			Uint64("revision", *req.Revision).
+			Msg("revision provided to begin stream")
+
+		esSubscribeToAllOptions.From = esdb.Position{
+			Commit:  *req.Revision,
+			Prepare: *req.Revision,
+		}
+	}
+
+	esStream, err := s.es.SubscribeToAll(ctx, esSubscribeToAllOptions)
+	if err != nil {
+		return err
+	}
+	defer esStream.Close()
+
+	for {
+		esdbEvent := esStream.Recv()
+
+		select {
+		case <-stream.Context().Done():
+			return nil
+		default:
+		}
+
+		if esdbEvent.SubscriptionDropped != nil {
+			esStream.Close()
+			log.Debug().Msg("subscription dropped, close stream")
```

Comment:
does esdbEvent.SubscriptionDropped supply more information, which may be interesting to be logged?